### DISCUSSION TITLE
Persistent TUI

### DISCUSSION THREAD

**Discussion by lishaduck** [vercel/turborepo#8574] [Ideas]
### Goals

1. Make it easier to see errors

### Non-goals

1. Support all run-pty features out of the box

### Background

I use turbo as a better [run-pty](https://github.com/lydell/run-pty).

### Proposal

The main thing I've missed since switching to turborepo is the ability to see errors in the TUI. Once there's an error, turborepo quits. Now that's a TUI, would it be possible to keep it open rather than printing them? I'd like to keep the TUI open after an error, and ideally be able to restart failed tasks if they just flaked or I've fixed them. It's jarring when I'm inspecting a result or status and then something fails. It would also be nice if the other tasks kept running, but that's probably out of scope for this particular request.

I don't know Rust, but this does bother me enough I'd be willing to give it a shot sometime if y'all decide this is worthwhile.

**Comment by anthonyshew** [comment]
Do you have a reproduction of this case you could share?

**Reply by lishaduck** [reply]
I don't really get what you mean, but here's an example:

<img width="940" alt="Screenshot 2024-06-24 at 4 56 13 PM" src="https://github.com/vercel/turbo/assets/88557639/5c50d1f4-3176-446d-bd09-d58167468107">
Essentially, I'd like to stay in the TUI, rather than printing this out.

**Reply by lishaduck** [reply]
#8570 seems like it'll help, but I'd still prefer remaining in the TUI.

**Reply by anthonyshew** [reply]
It sounds like you might want `--continue` for this?

**Reply by lishaduck** [reply]
That keeps the other tasks running, but once they finish, it'll quit the TUI. I'd prefer to stay in it even once its finished.

**Reply by anthonyshew** [reply]
Re-reading again, it's sounding like you're looking for Watch Mode: https://turbo.build/repo/docs/crafting-your-repository/developing-applications#watch-mode

This would mean a change to the errored package or app or its dependencies would get a retry when you edit source code and the whole of the `turbo` process will be kept open.

**Reply by lishaduck** [reply]
I'll try continue with watch when I get some time, and see if it's what I'm looking for.
Thanks!

**Reply by lishaduck** [reply]
Oh, it looks like `turbo watch` doesn't support single project repositories.

**Comment by wassim-ben-amor** [comment]
I have the same concern 🙏 
https://github.com/vercel/turbo/discussions/8566
I think it will be very useful to have an option to disable the terminal UI auto close

**Reply by anthonyshew** [reply]
Have you tried `turbo watch`?

**Reply by wassim-ben-amor** [reply]
`turbo watch` is always triggering new run without any change .. I didn't find a way to stop the retries :(

**Reply by anthonyshew** [reply]
This is likely because something is being written to disk but is not gitignored. Does this sound familiar? https://turbo.build/repo/docs/reference/watch#task-outputs

**Reply by wassim-ben-amor** [reply]
I can't see something written ..
Do you know if there is a way to log what have been changed ?
Thank you,

**Comment by felipedeboni** [comment]
@anthonyshew is there a way to keep TUI open after it's done? Tried a bunch of flags and I can't keep it open when it finishes for all packages.

It would be awesome to a flag for it :)

**Reply by lishaduck** [reply]
There doesn't seem to be.
Turborepo makes the assumption that you'll want to automatically restart tasks if you wish to keep them open (and that you're in a monorepo)

**Reply by anthonyshew** [reply]
😄 https://github.com/vercel/turbo/discussions/8574#discussioncomment-9899583
